docs/source/nlp/language_modeling.rst

```diff
@@ -3,4 +3,170 @@
 Language Modeling
 =================
 
-An example script on how to train the model can be found here: `NeMo/examples/nlp/language_modeling <https://github.com/NVIDIA/NeMo/tree/main/examples/nlp/language_modeling>`__.
+A language model (LM) estimates the joint probability of a given text corpus :math:`(x_1,\dots,x_T)` by factorizing it with a chain rule :math:`P(x_1,\dots,x_T) = \prod_{t=1}^T P(x_t|x_1,\dots,x_{t-1})` and sequentially modeling each conditional term in the product. To simplify modeling, it is often assumed that the context size (a number of preceding words) necessary to predict each word :math:`x_t` in the corpus is limited to :math:`N:\;P(x_t|x_1,\dots,x_{t-1}) \approx P(x_t|x_{t-N},\dots,x_{t-1})`. This approximation is commonly referred to as N-gram LM.
+
+Currently, we mainly support sentence-level LMs which do not consider long-term dependencies and model all sentences independently of each other. Our models are based on the Transformer sequence-to-sequence architecture.
+
+An example script on how to train the model can be found here: `NeMo/examples/nlp/language_modeling/transformer_lm.py <https://github.com/NVIDIA/NeMo/tree/main/examples/nlp/language_modeling/transformer_lm.py>`_. The default configuration file for the model can be found at: `NeMo/examples/nlp/language_modeling/conf/transformer_lm_config.yaml <https://github.com/NVIDIA/NeMo/tree/main/examples/nlp/language_modeling/conf/transformer_lm_config.yaml>`_.
+
+
+Data Format
+-----------
+
+Unsupervised LMs require the corpus which comprises many examples of sentences from a particular domain (Wikipedia, news, Pubmed abstracts, etc). We assume that the data is formatted as a text file where each line corresponds to a separate sentence:
+
+.. list-table:: *Sentence-level LM coprus*
+   :widths: 20
+   :header-rows: 1
+
+   * - train.librispeech.txt
+   * - in a silver cake basket as the panins had at their party
+   * - let us pretermit that long comparison
+   * - poverty contempt and sickness treading on my heels i easily resolve not to be affrighted
+   
+It is common practice to apply data cleaning, normalization, and tokenization to the data prior to training LM and 
+NeMo expects already cleaned, normalized, and tokenized data. The only data pre-processing NeMo does is subword tokenization with BPE.
+
+.. note::
+    If LM is intended to be used in a conjunction with another model (e.g. re-scoring of ASR, shallow fusion with NMT), make sure that the training data is preprocessed accordingly (lower-case no punctuation for ASR, Moses tokenization/normalization for NMT). Otherwise, it might introduce inadequate LM scores.
+
+
+Training a BPE Tokenization
```

Comment:
Change title to a more general term like "Tokenizer Training"
How about use the "sub-word level" instead of BPE? We support different type of tokenizers, even char-level with sentence-piece, am I right? We can mention it here.